src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -0,0 +1,296 @@
+            var totalGlobalProperties = _projectCollection.GlobalProperties.Count + globalProperties.Count;
+
+            foreach (var loadedProject in loadedProjects)
+            {
+                if (loadedProject.GlobalProperties.Count != totalGlobalProperties)
+                {
+                    continue;
+                }
+
+                // All projects in the collection should have the default global properties, so
+                // there's no need to check collection.GlobalProperties.
+
+                var found = true;
+                foreach (var globalProp in globalProperties)
+                {
+                    if (!loadedProject.GlobalProperties.Contains(globalProp))
+                    {
+                        found = false;
+                        break;
+                    }
+                }
+
+                if (found)
+                {
+                    return loadedProject;
+                }
+            }
+
+            return null;
+        }
+
+        public async Task<(MSB.Evaluation.Project project, DiagnosticLog log)> LoadProjectAsync(
+            string path, IDictionary<string, string> globalProperties, CancellationToken cancellationToken)
+        {
+            var log = new DiagnosticLog();
+
+            try
+            {
+                var project = FindProject(path, globalProperties);
+                if (project != null)
+                {
+                    return (project, log);
+                }
+
+                using (var stream = await ReadFileAsync(path, cancellationToken).ConfigureAwait(false))
+                using (var xmlReader = XmlReader.Create(stream, s_xmlReaderSettings))
+                {
+                    var xml = MSB.Construction.ProjectRootElement.Create(xmlReader, _projectCollection);
+
+                    // When constructing a project from an XmlReader, MSBuild cannot determine the project file path.  Setting the
+                    // path explicitly is necessary so that the reserved properties like $(MSBuildProjectDirectory) will work.
+                    xml.FullPath = path;
+
+                    project = new MSB.Evaluation.Project(xml, globalProperties, toolsVersion: null, _projectCollection);
+
+                    return (project, log);
+                }
+            }
+            catch (Exception e)
+            {
+                log.Add(e, path);
+                return (project: null, log);
+            }
+        }
+
+        private static async Task<MemoryStream> ReadFileAsync(string path, CancellationToken cancellationToken)
+        {
+            var buffer = new byte[1024];
+
+            using (var stream = FileUtilities.OpenAsyncRead(path))
+            {
+                var totalBytesRead = 0;
+
+                int bytesRead;
+                while ((bytesRead = await stream.ReadAsync(buffer, totalBytesRead, buffer.Length - totalBytesRead, cancellationToken).ConfigureAwait(false)) > 0)
```

Comment:
we have this - http://index/?leftProject=Microsoft.CodeAnalysis.Workspaces&leftSymbol=agjkvgvdetph&file=Utilities%5CSerializableBytes.cs&line=133

that could make this simpler

Reply:
Will take a look -- thanks! The original MSBuildWorkspace already had code like this (though I've refactored it considerably).